File: backend/controllers/hospitals_controller.py
import requests
from flask import Blueprint, request, jsonify
import math

hospitals_bp = Blueprint('hospitals_bp', __name__)

def calculate_distance(lat1, lon1, lat2, lon2):
    """Calculate distance between two coordinates using Haversine formula"""
    R = 6371
    d_lat = (lat2 - lat1) * math.pi / 180
    d_lon = (lon2 - lon1) * math.pi / 180
    a = (
        math.sin(d_lat / 2) * math.sin(d_lat / 2) +
        math.cos(lat1 * math.pi / 180) * math.cos(lat2 * math.pi / 180) *
        math.sin(d_lon / 2) * math.sin(d_lon / 2)
    )
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    distance = R * c
    return round(distance, 2)

def estimate_eta(distance_km):
    """Estimate ETA based on distance"""
    avg_speed = 40
    minutes = round((distance_km / avg_speed) * 60)
    hours = minutes // 60
    mins = minutes % 60
    return f"{hours}h {mins}m" if hours > 0 else f"{mins}m"

@hospitals_bp.route('/api/hospitals/nearby', methods=['GET'])
def get_nearby_hospitals():
    try:
        lat = request.args.get('lat')
        lng = request.args.get('lng')
        radius = request.args.get('radius', 5000, type=int)
        
        if not lat or not lng:
            return jsonify({'error': 'Latitude and longitude are required'}), 400
        
        try:
            lat = float(lat)
            lng = float(lng)
        except ValueError:
            return jsonify({'error': 'Latitude and longitude must be valid numbers'}), 400
        
        radius_km = radius / 1000
        
        overpass_query = f"""
        [bbox:{float(lng) - (radius_km / 111)},{float(lat) - (radius_km / 111)},{float(lng) + (radius_km / 111)},{float(lat) + (radius_km / 111)}];
        (
          node["amenity"="hospital"];
          way["amenity"="hospital"];
          relation["amenity"="hospital"];
        );
        out center;
        """
        
        response = requests.post(
            'https://overpass-api.de/api/interpreter',
            data=overpass_query,
            headers={'Content-Type': 'application/x-www-form-urlencoded'},
            timeout=15
        )
        
        if response.status_code != 200:
            return jsonify({
                'success': True,
                'count': 0,
                'message': 'No hospitals found',
                'data': []
            }), 200
        
        data = response.json()
        
        if not data.get('elements'):
            return jsonify({
                'success': True,
                'count': 0,
                'message': 'No hospitals found in the specified radius',
                'data': []
            }), 200
        
        hospitals = []
        for element in data['elements']:
            if element.get('center') or (element.get('lat') and element.get('lon')):
                hosp_lat = element.get('center', {}).get('lat') or element.get('lat')
                hosp_lon = element.get('center', {}).get('lon') or element.get('lon')
                
                distance = calculate_distance(lat, lng, hosp_lat, hosp_lon)
                tags = element.get('tags', {})
                
                hospitals.append({
                    'name': tags.get('name', 'Unknown Hospital'),
                    'address': f"{tags.get('addr:street', '')}, {tags.get('addr:city', '')}".strip(', '),
                    'distance': f"{distance} km",
                    'eta': estimate_eta(distance),
                    'coordinates': {
                        'lat': hosp_lat,
                        'lng': hosp_lon
                    },
                    'type': tags.get('healthcare:speciality', 'General Hospital'),
                    'phone': tags.get('contact:phone', 'N/A'),
                    'website': tags.get('website', 'N/A'),
                    'beds': tags.get('beds:number', 'Not available')
                })
        
        hospitals.sort(key=lambda x: float(x['distance'].split()[0]))
        
        return jsonify({
            'success': True,
            'count': len(hospitals),
            'userLocation': {'lat': lat, 'lng': lng},
            'data': hospitals,
            'note': 'Data from OpenStreetMap (free & open source)'
        }), 200
        
    except requests.exceptions.Timeout:
        return jsonify({
            'success': True,
            'count': 0,
            'message': 'Request timeout - please try again',
            'data': []
        }), 200
    except Exception as e:
        return jsonify({
            'error': 'Failed to fetch hospitals',
            'details': str(e)
        }), 500

@hospitals_bp.route('/api/hospitals/ambulances/nearby', methods=['GET'])
def get_nearby_ambulances():
    try:
        lat = request.args.get('lat')
        lng = request.args.get('lng')
        radius = request.args.get('radius', 2000, type=int)
        
        if not lat or not lng:
            return jsonify({'error': 'Latitude and longitude are required'}), 400
        
        try:
            lat = float(lat)
            lng = float(lng)
        except ValueError:
            return jsonify({'error': 'Latitude and longitude must be valid numbers'}), 400
        
        radius_km = radius / 1000
        
        overpass_query = f"""
        [bbox:{float(lng) - (radius_km / 111)},{float(lat) - (radius_km / 111)},{float(lng) + (radius_km / 111)},{float(lat) + (radius_km / 111)}];
        (
          node["amenity"="ambulance_station"];
          way["amenity"="ambulance_station"];
          relation["amenity"="ambulance_station"];
        );
        out center;
        """
        
        response = requests.post(
            'https://overpass-api.de/api/interpreter',
            data=overpass_query,
            headers={'Content-Type': 'application/x-www-form-urlencoded'},
            timeout=15
        )
        
        if response.status_code != 200:
            return jsonify({
                'success': True,
                'count': 0,
                'message': 'No ambulance stations found',
                'data': []
            }), 200
        
        data = response.json()
        
        if not data.get('elements'):
            return jsonify({
                'success': True,
                'count': 0,
                'message': 'No ambulance stations found in the specified radius',
                'data': []
            }), 200
        
        ambulances = []
        for element in data['elements']:
            if element.get('center') or (element.get('lat') and element.get('lon')):
                amb_lat = element.get('center', {}).get('lat') or element.get('lat')
                amb_lon = element.get('center', {}).get('lon') or element.get('lon')
                
                distance = calculate_distance(lat, lng, amb_lat, amb_lon)
                tags = element.get('tags', {})
                
                ambulances.append({
                    'name': tags.get('name', 'Ambulance Station'),
                    'address': f"{tags.get('addr:street', '')}, {tags.get('addr:city', '')}".strip(', '),
                    'distance': f"{distance} km",
                    'eta': estimate_eta(distance),
                    'coordinates': {
                        'lat': amb_lat,
                        'lng': amb_lon
                    },
                    'phone': tags.get('contact:phone', 'N/A'),
                    'operator': tags.get('operator', 'Not specified')
                })
        
        ambulances.sort(key=lambda x: float(x['distance'].split()[0]))
        
        return jsonify({
            'success': True,
            'count': len(ambulances),
            'data': ambulances,
            'note': 'Data from OpenStreetMap (free & open source)'
        }), 200
        
    except requests.exceptions.Timeout:
        return jsonify({
            'success': True,
            'count': 0,
            'message': 'Request timeout - please try again',
            'data': []
        }), 200
    except Exception as e:
        return jsonify({
            'error': 'Failed to fetch ambulances',
            'details': str(e)
        }), 500

File: backend/controllers/hospitalsController.js
import axios from 'axios';

const calculateDistance = (lat1, lon1, lat2, lon2) => {
  const R = 6371;
  const dLat = ((lat2 - lat1) * Math.PI) / 180;
  const dLon = ((lon2 - lon1) * Math.PI) / 180;
  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos((lat1 * Math.PI) / 180) *
      Math.cos((lat2 * Math.PI) / 180) *
      Math.sin(dLon / 2) *
      Math.sin(dLon / 2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  const distance = R * c;
  return distance.toFixed(2);
};

const estimateETA = (distanceKm) => {
  const avgSpeed = 40;
  const minutes = Math.round((distanceKm / avgSpeed) * 60);
  const hours = Math.floor(minutes / 60);
  const mins = minutes % 60;
  return hours > 0 ? `${hours}h ${mins}m` : `${mins}m`;
};

export const getNearbyHospitals = async (req, res, next) => {
  try {
    const { lat, lng, radius = 5000 } = req.query;

    if (!lat || !lng) {
      return res.status(400).json({
        error: 'Latitude and longitude are required'
      });
    }

    if (isNaN(lat) || isNaN(lng)) {
      return res.status(400).json({
        error: 'Latitude and longitude must be valid numbers'
      });
    }

    const radiusKm = parseInt(radius) / 1000;
    const overpassQuery = `
      [bbox:${parseFloat(lng) - (radiusKm / 111)},${parseFloat(lat) - (radiusKm / 111)},${parseFloat(lng) + (radiusKm / 111)},${parseFloat(lat) + (radiusKm / 111)}];
      (
        node["amenity"="hospital"];
        way["amenity"="hospital"];
        relation["amenity"="hospital"];
      );
      out center;
    `;

    const response = await axios.post('https://overpass-api.de/api/interpreter', overpassQuery, {
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' }
    });

    if (!response.data.elements || response.data.elements.length === 0) {
      return res.status(404).json({
        success: true,
        count: 0,
        message: 'No hospitals found in the specified radius',
        data: []
      });
    }

    const hospitals = response.data.elements
      .filter(el => el.center || (el.lat && el.lon))
      .map((hospital) => {
        const hospitalLat = hospital.center?.lat || hospital.lat;
        const hospitalLon = hospital.center?.lon || hospital.lon;
        const distance = calculateDistance(
          parseFloat(lat),
          parseFloat(lng),
          hospitalLat,
          hospitalLon
        );

        return {
          name: hospital.tags?.name || 'Unknown Hospital',
          address: hospital.tags?.['addr:street'] 
            ? `${hospital.tags['addr:street']}, ${hospital.tags['addr:city'] || ''}`
            : 'Address not available',
          distance: `${distance} km`,
          eta: estimateETA(distance),
          coordinates: {
            lat: hospitalLat,
            lng: hospitalLon
          },
          type: hospital.tags?.['healthcare:speciality'] || 'General Hospital',
          phone: hospital.tags?.['contact:phone'] || 'N/A',
          website: hospital.tags?.website || 'N/A',
          operatorType: hospital.tags?.operator || 'Not specified',
          beds: hospital.tags?.['beds:number'] || 'Not available'
        };
      })
      .sort((a, b) => parseFloat(a.distance) - parseFloat(b.distance));

    res.status(200).json({
      success: true,
      count: hospitals.length,
      userLocation: {
        lat: parseFloat(lat),
        lng: parseFloat(lng)
      },
      data: hospitals,
      note: 'Data from OpenStreetMap (free & open source)'
    });
  } catch (error) {
    next(error);
  }
};

export const getHospitalDetails = async (req, res, next) => {
  try {
    const { lat, lng, name } = req.query;

    if (!lat || !lng || !name) {
      return res.status(400).json({
        error: 'lat, lng, and name are required'
      });
    }

    const overpassQuery = `
      [bbox:${parseFloat(lng) - 0.01},${parseFloat(lat) - 0.01},${parseFloat(lng) + 0.01},${parseFloat(lat) + 0.01}];
      (
        node["amenity"="hospital"]["name"="${name}"];
        way["amenity"="hospital"]["name"="${name}"];
        relation["amenity"="hospital"]["name"="${name}"];
      );
      out center;
    `;

    const response = await axios.post('https://overpass-api.de/api/interpreter', overpassQuery, {
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' }
    });

    if (!response.data.elements || response.data.elements.length === 0) {
      return res.status(404).json({
        error: 'Hospital details not found'
      });
    }

    const hospital = response.data.elements[0];

    res.status(200).json({
      success: true,
      data: {
        name: hospital.tags?.name || 'Unknown',
        address: `${hospital.tags?.['addr:street'] || ''}, ${hospital.tags?.['addr:city'] || ''}`,
        phone: hospital.tags?.['contact:phone'] || 'N/A',
        website: hospital.tags?.website || 'N/A',
        type: hospital.tags?.['healthcare:speciality'] || 'General Hospital',
        operatorType: hospital.tags?.operator || 'Not specified',
        beds: hospital.tags?.['beds:number'] || 'Not available',
        coordinates: {
          lat: hospital.center?.lat || hospital.lat,
          lng: hospital.center?.lon || hospital.lon
        }
      }
    });
  } catch (error) {
    next(error);
  }
};

export const getNearbyAmbulances = async (req, res, next) => {
  try {
    const { lat, lng, radius = 2000 } = req.query;

    if (!lat || !lng) {
      return res.status(400).json({
        error: 'Latitude and longitude are required'
      });
    }

    const radiusKm = parseInt(radius) / 1000;
    const overpassQuery = `
      [bbox:${parseFloat(lng) - (radiusKm / 111)},${parseFloat(lat) - (radiusKm / 111)},${parseFloat(lng) + (radiusKm / 111)},${parseFloat(lat) + (radiusKm / 111)}];
      (
        node["amenity"="ambulance_station"];
        way["amenity"="ambulance_station"];
        relation["amenity"="ambulance_station"];
      );
      out center;
    `;

    const response = await axios.post('https://overpass-api.de/api/interpreter', overpassQuery, {
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' }
    });

    if (!response.data.elements || response.data.elements.length === 0) {
      return res.status(404).json({
        success: true,
        count: 0,
        message: 'No ambulance stations found in the specified radius',
        data: []
      });
    }

    const ambulances = response.data.elements
      .filter(el => el.center || (el.lat && el.lon))
      .map((ambulance) => {
        const ambLat = ambulance.center?.lat || ambulance.lat;
        const ambLon = ambulance.center?.lon || ambulance.lon;
        const distance = calculateDistance(
          parseFloat(lat),
          parseFloat(lng),
          ambLat,
          ambLon
        );

        return {
          name: ambulance.tags?.name || 'Ambulance Station',
          address: ambulance.tags?.['addr:street'] 
            ? `${ambulance.tags['addr:street']}, ${ambulance.tags['addr:city'] || ''}`
            : 'Address not available',
          distance: `${distance} km`,
          eta: estimateETA(distance),
          coordinates: {
            lat: ambLat,
            lng: ambLon
          },
          phone: ambulance.tags?.['contact:phone'] || 'N/A',
          operator: ambulance.tags?.operator || 'Not specified'
        };
      })
      .sort((a, b) => parseFloat(a.distance) - parseFloat(b.distance));

    res.status(200).json({
      success: true,
      count: ambulances.length,
      data: ambulances,
      note: 'Data from OpenStreetMap (free & open source)'
    });
  } catch (error) {
    next(error);
  }
};
